File: translated/f-to-c/DRB017-outputdep-var-yes.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//The loop in this example cannot be parallelized.
//
//Data race pairs: we allow two pairs to preserve the original code pattern.
// 1. x@50:16:R vs. x@51:9:W
// 2. x@51:9:W vs. x@51:9:W

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char* argv[])
{
    int len, i, x;
    int* a;

    len = 100;
    x = 10;

    if (argc == 1) {
        printf("No command line arguments provided.\n");
    }

    if (argc >= 2) {
        len = atoi(argv[1]);
        if (len <= 0) {
            printf("Error, invalid integer value.\n");
        }
    }

    a = (int*)malloc(len * sizeof(int));

    #pragma omp parallel for
    for (i = 0; i < len; i++) {
        a[i] = x;
        x = i + 1;  // Adjust for 0-based indexing in C
    }

    printf("x = %d  a(0) = %d\n", x, a[0]);

    free(a);

    return 0;
} 

File: translated/f-to-go/DRB017-outputdep-var-yes.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//The loop in this example cannot be parallelized.
//
//Data race pairs: we allow two pairs to preserve the original code pattern.
// 1. x@50:16:R vs. x@51:9:W
// 2. x@51:9:W vs. x@51:9:W

package main

import (
	"fmt"
	"os"
	"runtime"
	"strconv"
	"sync"
)

func main() {
	var length, x, argCount, allocStatus, ix int
	var rdErr error
	var args []string
	var a []int

	length = 100
	x = 10

	argCount = len(os.Args) - 1
	if argCount == 0 {
		fmt.Printf("No command line arguments provided.\n")
	}

	args = make([]string, argCount)
	allocStatus = 0
	if allocStatus > 0 {
		fmt.Printf("Allocation error, program terminated.\n")
		os.Exit(1)
	}

	for ix = 1; ix <= argCount; ix++ {
		if ix <= len(os.Args)-1 {
			args[ix-1] = os.Args[ix]
		}
	}

	if argCount >= 1 {
		length, rdErr = strconv.Atoi(args[0])
		if rdErr != nil {
			fmt.Printf("Error, invalid integer value.\n")
		}
	}

	a = make([]int, length)
	//$omp parallel do
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := length / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= length; start += chunkSize {
		end := start + chunkSize - 1
		if end > length {
			end = length
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				a[i-1] = x // Race condition: reading x
				x = i      // Race condition: writing x
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do

	// Handle a(0) access - in Fortran this is out of bounds, in Go it would be a[-1]
	var a0 int
	if length > 0 {
		a0 = a[0] // This represents a(1) in Fortran, but original tries to access a(0)
	}
	fmt.Printf("x=%3d  a(0)=%3d\n", x, a0)

	// deallocate(args,a)
}
